llvm/docs/LangRef.rst

```diff
@@ -19234,6 +19234,78 @@ the follow sequence of operations:
 
 The ``mask`` operand will apply to at least the gather and scatter operations.
 
+
+.. _int_masked_compress:
+
+'``llvm.masked.compress.*``' Intrinsics
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+LLVM provides an intrinsic for compressing data within a vector based on a selection mask.
+Semantically, this is similar to :ref:`llvm.masked.compressstore <int_compressstore>` but with weaker assumptions
+and without storing the results to memory, i.e., the data remains in the vector.
+
+Syntax:
+"""""""
+This is an overloaded intrinsic. A number of scalar values of integer, floating point or pointer data type are collected
+from an input vector and placed adjacently within the result vector. A mask defines which elements to collect from the vector.
+The remaining lanes are filled with values from ``passthru``.
+
+:: code-block:: llvm
+
+      declare <8 x i32> @llvm.masked.compress.v8i32(<8 x i32> <value>, <8 x i1> <mask>, <8 x i32> <passthru>)
+      declare <16 x float> @llvm.masked.compress.v16f32(<16 x float> <value>, <16 x i1> <mask>, <16 x float> undef)
+
+Overview:
+"""""""""
+
+Selects elements from input vector '``value``' according to the '``mask``'.
+All selected elements are written into adjacent lanes in the result vector, from lower to higher.
+The mask holds an entry for each vector lane, and is used to select elements to be kept.
+If ``passthru`` is undefined, the number of valid lanes is equal to the number of ``true`` entries in the mask, i.e., all lanes >= number-of-selected-values are undefined.
+If a ``passthru`` vector is given, all remaining lanes are filled with the corresponding lane's value from ``passthru``.
+The main difference to :ref:`llvm.masked.compressstore <int_compressstore>` is that the we do not need to guard against memory access for unselected lanes.
```

Comment:
Please wrap lines at 80 columns

Probably worth explicitly stating what happens if the mask has poison/undef bits.  (And make sure the expansion freezes store indexes as necessary to implement those semantics.)

Reply:
Good point, I've added this in the docs. I would interpret `undef` as "false" (this how I implemented it), because we cannot access that element and the caller says "I don't care about this value". I guess this does not need any freezing stores, as we handle this deterministically.

Reply:
That doesn't work.  There's no way at runtime to compute whether an input is undef/poison at runtime.  (You can technically special-case a literal undef/poison, but that doesn't really do anything useful.)

Reply:
Ah, okay. I think my understanding of when a freeze is needed was wrong. In this case, I'm not sure we can specify one or the other behavior if a mask element is undef/poison. Depending on the code path, we perform a vec_reduce to get a passthru value or not. So our result depends on how vec_reduce behaves with undef/poison in addition to how our extracted mask element if frozen, at which point this probably gets weird to handle consistently.

Before changing the docs again: Is it valid for us to say "If the mask is undef or contains undef elements, the entire result is undef"? That would give us a bit of wiggle room in the implementation.

Reply:
If a mask element is poison, I think it's fine if the whole returned vector is poison.  But preferably not undefined behavior; undefined behavior means we can't hoist, which makes everything more complicated, so I'd rather not deal with that.  (I guess you could try to more narrowly state that elements after the poisoned mask bit are poison, but I don't think that's actually useful.)

If a mask element is undef... I guess we could similarly say the whole returned vector is undef?  Not sure how much it matters exactly what rule we pick here.

Reply:
What does `@llvm.masked.compressstore` currently do with undef or poison mask elements? It would be nice if they were consistent

Reply:
@lukel97  I just played around a bit with `@llvm.masked.compressstore` and immediate values. It is does not seem very consistent (https://godbolt.org/z/xjeY5z9bb). Sometimes and `undef` value is treated as true, sometimes as false. I'm not sure what the takeaway from this is though.

@efriedma-quic Sorry, but I'm a bit lost in your distinction here. Could you maybe suggest some wording on how to treat this? I think we basically want to say that if a mask is/contains poison/undef bits, we don't care about the result vector. I just don't know how to express this with the correct LLVM terminology.

Reply:
Strawman wording: "if any element of the mask is `poison`, all elements of the result are `poison`.  Otherwise, if any element of the mask is `undef`, all elements of the result are `undef`."

This mostly only affects optimizations you haven't implemented yet... but probably the code to expand llvm.masked.compress to stores needs to freeze the mask, so we don't do a store with a poisoned pointer operand.